Add a prosaic introduction to distributed data structures (#6243)
microsoft/FluidFramework: docs/content/docs/data-structures/overview.md

---
title: Overview of Distributed Data Structures
menuPosition: 1
---

The Fluid Framework provides developers with distributed data structures (DDSes) that automatically ensure that each

> connected client has access to the same state. The APIs provided by DDSes are designed to be familiar to programmers

Reviewer: > each connected client has access to the same state
**eventually**, right? Otherwise it reads as if every operation is blocking.

Author: Good point

> who’ve used common data structures before.

Reviewer 2: dumb quotes please

Reviewer 2: here and elsewhere

Author: What does this mean?

Reviewer 2: ```suggestion
who've used common data structures before.
```

Sorry, I wasn't clear. Does the code change clarify? I'll look into a lint rule for this.

> When using a DDS, you can largely treat it as a local object. You can add data to it, remove data, update it, etc.

Reviewer: There are a lot of caveats with this statement. Related data across DDSs, or even guidance below to split content into multiple keys within same DDS requires people to consider impacts of concurrent edits. I feel we need some write up on that, and a reference to it every time we say it's as simple as local object, as we see even our own engineers continue to make mistakes in this space.

Reviewer 2: I'm responsible for the original wording, and I approve Vlad's message. :)

Reviewer: BTW, I'm happy to write article / docs (stand-alone :)), as long as you promise to help to prof-read and integrate into right place :) Especially on concurrency

Author: That'd be awesome!

However, a DDS is not just a local object. A DDS can also be changed by other users that are editing.

Choosing the correct data structure for your scenario can improve the performance and code structure of your application.
Below we've enumerated the data structures and described when they may be most useful.

## Key-value Data

These DDSes are used for storing key-value data. They are optimistic and use a last-writer-wins merge policy.

* [SharedMap]({{< relref "/docs/apis/map/sharedmap.md" >}}) - a basic key-value data structure
* [SharedDirectory]({{< relref "/docs/apis/map/shareddirectory.md" >}}) – a SharedMap with hierarchical paths instead of
simple keys
* [SharedCell]({{< relref "/docs/apis/cell/sharedcell.md" >}}) – a “single-object SharedMap”; useful for wrapping objects.

### Key Value Scenarios

Key-value data structures are the most common choice for many scenarios.

* user preference data
* current state of a survey
* the configuration of a view

### Common Issues

> * Storing a counter in a map will have unexpected behavior. Use the SharedCounter instead.

Reviewer: I'd replace "unexpected behavior" with something concrete. I.e. you can't atomically / safely increment it

Reviewer 2: If you can't expand on this sufficiently here, which is likely, then you could link to another unwritten article. Not ideal but a reasonable approach if needed.

Author: Good call out

> * Storing a lot of data in one key-value entry may cause performance issues. Each update will update the entire value.

Reviewer: Perf is important, but I feel like that's not the primary reason for guidance - it's the atomicity / scoping, i.e. what is changed as one atomic unit. There is no merging of sub-parts of object stored in the key.

Try splitting the data across multiple keys
* Storing arrays, lists, or logs in a key-value entry may lead to unexpected behavior because users can't collaboratively
modify parts of one entry. Try storing the array or list data in a SharedSequence or SharedInk

## Sequences

These DDSes are used for storing sequential data. They are optimistic.

* [SharedNumberSequence]({{< relref "SharedNumberSequence" >}}) – a sequence of numbers.

> * [SharedObjectSequence]({{< relref "/docs/apis/sequence/sharedobjectsequence.md" >}}) – a sequence of objects.

Reviewer 2: "plain objects"?

Author: 👍

* [SharedMatrix]({{< relref "SharedMatrix" >}}) – a data structure to efficiently use two-dimensional tabular data.
* [SharedString]({{< relref "SharedString" >}}) – a specialized data structure for handling collaborative text.

> Sequence Data Structures are useful when you'll need to insert data in the middle of a list or array. Unlike the

Reviewer 2: ```suggestion
Sequence data structures are useful when you'll need to insert data in the middle of a list or array. Unlike the
```

key-value data structures, sequences have a sequential order and can handle simultaneous inserts from multiple users.

### Sequence Scenarios

* Text editors

> * Tabular data

Reviewer 2: Why not SharedMatrix?

Author: SharedMatrix is a sequence scenario?

Reviewer 2: Well, when I think tabular data, I don't think sequence. Wouldn't SharedMatrix be a better fit for tabular data?

* Timelines
* Lists

## Specialized data structures

* [SharedCounter]({{< relref "SharedCounter" >}}) – a counter.

    The SharedCounter is useful to keep track of increments. While a key-value data structure appears like a good fit,
    two users simultaneously setting the same key can cause issues.

* [Ink]({{< relref "/docs/apis/ink/ink.md" >}}) – a specialized data structure for ink data.

>     Ink is a specific form of an append only list. It's great for capturing ink strokes.

Reviewer 2: ```suggestion
    Ink is a form of an append only list. It's intended for capturing ink strokes.
```

## Consensus Data Structures

These DDSes are **not** optimistic. Before a change to a consensus data structure is confirmed, the connected clients
must acknowledge the change.

> * [OrderedCollection]({{< relref "/docs/apis/ordered-collection" >}}) - an ordered queue of items, but each item is pulled
> off the queue by only one client
> * [RegisterCollection]({{< relref "/docs/apis/register-collection" >}}) - Stores values, but keeps a record of all changes
> * Quorum - Allows clients to agree on a proposal. Quorum also contains client information

Reviewer 2: ```suggestion
* [OrderedCollection]({{< relref "/docs/apis/ordered-collection" >}}) -- an ordered queue of items, but each item is pulled
off the queue by only one client.
* [RegisterCollection]({{< relref "/docs/apis/register-collection" >}}) -- Stores values, but keeps a record of all changes.
* Quorum -- Allows clients to agree on a proposal. Quorum also contains client information.
```

Periods at the end of list items and `--` for an en-dash instead of hyphen.

Reviewer 2: here and elsewhere

### Consensus Scenarios

> People use consensus data structures to guarantee that only one client does an action, or that all clients consent to

Reviewer 2: ```suggestion
Consensus data structures guarantee that only one client does an action, or that all clients consent to
```
